Mathlib/Order/ReltoGaloisConnection.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2024 Lagrange Mathematics and Computing Research Center. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anthony Bordg
+-/
+import Mathlib.Data.Rel
+import Mathlib.Order.GaloisConnection
+
+/-!
+# The Galois Connection Induced by a Relation
+
+In this file, we show that an arbitrary relation `R` between a pair of types `α` and `β` defines
+a pair `leftDual` and `rightDual` of adjoint order-preserving maps between the corresponding posets
+`Set α` and `(Set β)ᵒᵈ`.
+We define `lFixedPoints` (resp. `rFixedPoints`) as the set of fixed points `J` (resp. `I`) of
+`Set α` (resp. `(Set β)ᵒᵈ`) such that `rightDual (leftDual J) = J` (resp.
+`leftDual (rightDual I) = I`).
+
+## Main Results
+
+⋆ `galoisConnection`: we prove that the maps `leftDual` and `rightDual` form a Galois connection.
+⋆ `equivFixedPoints`: we prove that the said maps induce inverse bijections between the sets
+of fixed points.
+
+## References
+
+⋆ Engendrement de topologies, démontrabilité et opérations sur les sous-topos, Olivia Caramello and
+Laurent Lafforgue (in preparation)
+
+## Tags
+
+relation, Galois connection, induced bijection, fixed points
+-/
+
+section
+
+universe u v
+
+variable {α : Type u} {β : Type v} (R : Rel α β)
+
+namespace Rel
+
+/-! ### Pairs of adjoint maps defined by relations -/
+
+open OrderDual
+
+/-- `leftDual` maps any set `J` of elements of type `α` to the set `{b : β | ∀ a ∈ J, R a b}` of
+elements `b` of type `β` such that `R a b` for every element `a` of `J`. -/
+def leftDual (J : Set α) : Set β := {b : β | ∀ ⦃a⦄, a ∈ J → R a b}
+
+/-- `rightDual` maps any set `I` of elements of type `β` to the set `{a : α | ∀ b ∈ ofDual I, R a b}`
+of elements `a` of type `α` such that `R a b` for every element `b` of `I`. -/
+def rightDual (I : Set β) : Set α := {a : α | ∀ ⦃b⦄, b ∈ I → R a b}
+
+/-- The pair of functions `leftDual` and `rightDual` forms a Galois connection. -/
+theorem gc_leftDual_rightDual : GaloisConnection (toDual ∘ R.leftDual) (R.rightDual ∘ ofDual) :=
+  fun J I ↦ ⟨fun h a ha b hb ↦ h (by simpa) ha, fun h b hb a ha ↦ h (by simpa) hb⟩
+
+/-! ### Induced equivalences and generation processes -/
+
+/-- `lFixedPoints` is the set of elements `J : Set α` satisfying `rightDual (leftDual J) = J`. -/
+def lFixedPoints := {J : Set α | R.rightDual (R.leftDual J) = J}
+
+/-- `rFixedPoints` is the set of elements `I : Set β` satisfying `leftDual (rightDual I) = I`. -/
+def rFixedPoints := {I : Set β | R.leftDual (R.rightDual I) = I}
+
+open GaloisConnection
+
+/-- `leftDual` maps every element `J` to `rFixedPoints`. -/
+theorem leftDual_mem_rFixedPoint (J : Set α) : R.leftDual J ∈ R.rFixedPoints := by
+  unfold rFixedPoints
+  apply le_antisymm
+  · apply R.gc_leftDual_rightDual.monotone_l; exact R.gc_leftDual_rightDual.le_u_l J
+  · exact R.gc_leftDual_rightDual.l_u_le (R.leftDual J)
```

Comment:
This can be turned into a proof term as follows:

```suggestion
theorem leftDual_mem_rFixedPoint (J : Set α) : R.leftDual J ∈ R.rFixedPoints :=
  le_antisymm (R.gc_leftDual_rightDual.monotone_l (R.gc_leftDual_rightDual.le_u_l J))
    (R.gc_leftDual_rightDual.l_u_le (R.leftDual J))
```

I have tested the code so you can safely commit this suggestion if you like.

Reply:
Are proof terms or tactics preferred in mathlib4, wrt legibility, maintenance etc?

Reply:
Really depends on the length. Here I think a tactic proof is justified

Reply:
As far as I know, there is no general convention in mathlib4 regarding the preference between proof terms and tactics. 

In my personal experience as an author (~ 7 months) and as a reviewer (~ 1 month), for relatively simple and short proofs such as these, fewer lines of code (and possibly faster execution) are typically considered better. 

Given a 5-10 lines simple proof (e.g. an intro and a sequence of apply), I think we can fairly say that a 1-2 lines proof term is better.

I don’t believe there is any significant loss in readability and maintainability in this case. 

Please @YaelDillies, correct me if I'm wrong.

Reply:
Ops, sorry @YaelDillies. For some reason your comment didn't show up on the mobile app...

Reply:
I think it's just not a big deal either way

Reply:
The only comment I have about these tactic blocks is that the `unfold` step is (most likely) unnecessary, and that's one tactic that we tend to avoid in mathlib proofs. @AnthonyBordg, could you try to get rid of these `unfold`s in your proofs?

Reply:
Done. Howewer, I have an error after the renaming the file `Order.Rel.GaloisConnection` as suggested by Yael.

Reply:
Seems like you forgot to run `lake exe mk_all`

Reply:
I did run `lake exe mk_all`.